ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/jca/TransactionImporterImple.java

```diff
@@ -122,6 +112,14 @@ public TransactionImple recoverTransaction(Uid actId)
 		if (recovered.baseXid() == null)
 		    throw new IllegalArgumentException();
 		
+		return addImportedTransaction(recovered);
+	}
+
+	private TransactionImple addImportedTransaction(TransactionImple importedTransaction) throws XAException {
+		SubordinateXidImple importedXid = new SubordinateXidImple(importedTransaction.baseXid());
+		TransactionImpleHolder holder = new TransactionImpleHolder();
+		TransactionImpleHolder prevHolder = _transactions.putIfAbsent(importedXid, holder);
+
```

Comment:
@tomjenkinson I reasoned that none of those putIfAbsent exceptions applied - besides if the put failed then it would show up as a RTE (ie no concurrency issues).

@dmlloyd I mean't that if some external agent interrupted the thread *after* the empty holder was added but before the setter was called then the setter would never get called.

But let me take a look at the suggestion that @dmlloyd gave regarding using a single point of synchronized initialization.